### ISSUE TITLE
Can't open my DAC as a 24/96 outstream on Linux

### ISSUE THREAD

**Issue by 0joshuaolson1** [andrewrk/libsoundio#94]
I'm not sure if libsoundio is to blame, but my external sound card is supposed to support 24bit/96kHz (Fiio E10). The C equivalent of what's failing is:
```
...
SoundIoOutStream *out = soundio_outstream_create(device); // not null for me
out->sample_rate = 96000;
out->format = SoundIoFormatS24LE;
int err = soundio_outstream_open(out)); // soundio_strerror is "unable to open device"

This is a raw device, Linux 4.5 and updated alsa 1.1.1. I get the same results with or without pulseaudio uninstalled. None of the 24-bit formats work and neither `device->formats` nor `soundio_device_supports_format` think it can do it either.

Should I contact the manufacturer or something? It's USB plug-and-play, so there are no drivers involved.

**Comment by andrewrk** [comment]
Is there another open source application on Linux that can open the device in raw mode with 96000 and it works? If so, we can cross reference the source code and see if we can spot a difference.

**Comment by 0joshuaolson1** [comment]
First I'm going to see if this happens in C as well. I was using Terra, and when I open in 16/48 I get a pthread segfault.

**Comment by 0joshuaolson1** [comment]
Okay, this segfaults on me after printing `Yay!` at the end (both release and debug libsoundio builds, `git pull`ed just now):
```
#include <soundio/soundio.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
int main(int argc, char **argv){
  struct SoundIo *sio = soundio_create();
  if(!sio){
    printf("out of memory\n");
    return 1;
  }
  int err = soundio_connect_backend(sio, SoundIoBackendAlsa);
  if(err){
    printf("%s\n", soundio_strerror(err));
    return 0;
  }
  soundio_flush_events(sio);
  struct SoundIoDevice *device;
  for(int i = 0; i < soundio_output_device_count(sio); i++){
    device = soundio_get_output_device(sio, i);
    if(strcmp(device->id, "hw:1,0") == 0)
      break;
    soundio_device_unref(device);
  }
  printf("%s\n", device->id);
  struct SoundIoOutStream *out = soundio_outstream_create(device);
  if(!out){
    printf("out of memory\n");
    return 1;
  }
  out->sample_rate = 48000; // or 96000
  out->format = SoundIoFormatS16LE; // the 16-bit format my device supports
  err = soundio_outstream_open(out);
  if(err){
    printf("can't open: %s\n", soundio_strerror(err));
    return 0;
  }
  if(out->layout_error){
    printf("can't set layout: %s\n", soundio_strerror(out->layout_error));
    return 0;
  }
  err = soundio_outstream_start(out);
  if(err){
    printf("can't start: %s\n", soundio_strerror(err));
    return 0;
  }
  printf("Yay!\n");
}
```
I put this code in place of example/sio_list_devices.c because CMake builds it for me. How do you suggest I debug?

**Comment by 0joshuaolson1** [comment]
> Is there another open source application on Linux that can open the device in raw mode with 96000 and it works? If so, we can cross reference the source code and see if we can spot a difference.

I'm installing a couple things to try.

**Comment by andrewrk** [comment]
Couple things in your example code:

 * after printf yay, do this:

```c
for (;;)
        soundio_wait_events(soundio);
```

Simple way to make your program block so that the audio loop can run.

 * You need to set the `write_callback` of the output stream. That's probably why it's segfaulting. Libsoundio currently doesn't assert or return an error code if you forget to set it, it just tries to call it.

> I put this code in place of example/sio_list_devices.c because CMake builds it for me. How do you suggest I debug?

That's fine, nothing wrong with that.

**Comment by 0joshuaolson1** [comment]
Great timing. I'll ignore the segfault then. Using the amazing script from https://lacocina.nl/detect-alsa-output-capabilities:
```
mpd-configure/alsa-capabilities -a 'hw:1,0' -s
 0) USB Audio Class Digital alsa audio output interface `hw:1,0'
 - device name       = DigiHug USB Audio                                           
 - interface name    = USB Audio                                                   
 - usb audio class   = 1 - isochronous adaptive                                    
 - character device  = /dev/snd/pcmC1D0p                                           
 - rates per format  = S24_3LE:             32000Hz 44100Hz 48000Hz 96000Hz        
                       S16_LE:              32000Hz 44100Hz 48000Hz 96000Hz        
 - monitor file      = /proc/asound/card1/pcm0p/sub0/hw_params                     
 - stream file       = /proc/asound/card1/stream0
```
My conclusion is libsoundio doesn't support `S24_3LE`.

**Comment by andrewrk** [comment]
> My conclusion is libsoundio doesn't support S24_3LE.

You are correct. I did a bit of checking, and it seems PulseAudio and ALSA do support this format, potentially Windows and CoreAudio too. So this is something that I will add to libsoundio. Thanks for this report.

**Comment by 0joshuaolson1** [comment]
You're welcome. Thanks for exposing the backend's raw powers instead of least denominator.

**Comment by andrewrk** [comment]
Sadly this is going to break ABI compatibility to add this format, so it has to bump major version.

**Comment by 0joshuaolson1** [comment]
Does this mean its priority is the same as the other 2.0.0 milestones? What would I need to do to patch it?

**Comment by andrewrk** [comment]
It's a reasonably straightforward patch to make. The trickiest part is going to be testing.

I think the path forward here, once we have a working patch for this, is to release a bug fix version with all the stuff we already have queued up. That'll be 1.1.1.

Then we can merge this patch in and release 2.0.0. No problem. The other 2.0.0 milestone items can be postponed. The only downside is it takes the various package managers a while to catch up to new major versions. I'm not too concerned about that, however. User using package managers can use the old version for a while, and developers who need the new stuff can temporarily build from source. (Or use a sweet package manager like NixOS which lets you override packages with new sources in a shell by shell basis)

To patch this involves finding out what the equivalent of S24_3LE is in the various audio backends, adding the new SoundIoFormat enum entries with a nice name and documentation, and hooking it up to each backend. Then testing which is the hard part.

**Comment by 0joshuaolson1** [comment]
Ugh, I don't know the equivalent of ALSA's [SND_PCM_FORMAT_S24_3LE](http://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ggaa14b7f26877a812acbb39811364177f8a1fd8c22f3dd45fc1b8e41b0b5e6107ce) - looking at output devices right now, I guess CoreAudio and WASAPI use more generic parameter combinations like block/container size for different bit depth deliveries (you may be familiar with this, but whatever):

https://developer.apple.com/library/prerelease/content/documentation/MusicAudio/Conceptual/CoreAudioOverview/CoreAudioEssentials/CoreAudioEssentials.html#//apple_ref/doc/uid/TP40003577-CH10-SW7
https://msdn.microsoft.com/en-us/library/dd797880(v=vs.85).aspx
https://msdn.microsoft.com/en-us/library/dd757714(v=vs.85).aspx

On Windows at least, there are standard formats but 8/16 bit only and the rest you have to guess by opening a stream:

https://msdn.microsoft.com/en-us/library/dd743855(v=vs.85).aspx

**Comment by andrewrk** [comment]
For PulseAudio at least: https://freedesktop.org/software/pulseaudio/doxygen/sample.html

```
PA_SAMPLE_S24LE - Signed 24 bit integer PCM packed, little endian.
PA_SAMPLE_S24BE - Signed 24 bit integer PCM packed, big endian.
```

**Comment by 0joshuaolson1** [comment]
For `SoundIoFormat(S|U)243(L|B)E`, should `soundio_get_bytes_per_sample` return `3` regardless of backend? The current 24-bit formats are designed to use native 32-bit machine data types.

**Comment by 0joshuaolson1** [comment]
(For future reference) Unlike the [sio_sine example](https://github.com/andrewrk/libsoundio/blob/master/example/sio_sine.c), write your samples similarly to ```soundIoChannelArea[channel].ptr[0 through 2] = (char)127;```, where little endian formats of course interpret the third byte (`ptr[2]`) as most significant.

The PulseAudio backend gives me underflows after a while, probably because I haven't configured it/my system for realtime priority? I don't see this problem with an identical Alsa callback.
